File: checkout-ruby/lib/promotional_rule_handler_factory.rb
class PromotionalRuleHandlerFactory

  def self.get_cart_promotional_list(promotional_rules)
    cart_promotional_list = []
    promotional_rules.each do |promotional_rule|
      if promotional_rule.get_discount_entity_type == RuleDiscountEntityType::CART
        cart_promotional_list << promotional_rule
      end
    end
    cart_promotional_list
  end


  def self.get_item_promotional_map(promotional_rules)
    item_promotional_map = {}
    promotional_rules.each do |promotional_rule|
      if promotional_rule.get_discount_entity_type == RuleDiscountEntityType::ITEM
        item_promotional_map[promotional_rule.item_code] = promotional_rule
      end
      item_promotional_map
    end
  end

end

File: src/main/java/com/vinterior/checkout/models/PromotionalRuleHandlerFactory.java
package com.vinterior.checkout.models;

import com.vinterior.checkout.enums.RuleDiscountEntityType;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class PromotionalRuleHandlerFactory {


    public static List<PromotionalRule> getCartPromoHandlers(List<PromotionalRule> promotionalRuleList) {
        List<PromotionalRule> cartPromotionalRuleList = new ArrayList<>();
        for(PromotionalRule promotionalRule: promotionalRuleList) {
            if (promotionalRule.getDiscountEntityType().equals(RuleDiscountEntityType.CART)) {
                cartPromotionalRuleList.add(promotionalRule);
            }
        }
        return cartPromotionalRuleList;
    }

    public static Map<Integer, PromotionalRule> getItemPromoHandlers(List<PromotionalRule> promotionalRuleList) {
        Map<Integer, PromotionalRule> itemPromotionalRuleMap = new HashMap<>();
        for(PromotionalRule promotionalRule: promotionalRuleList) {
            if (promotionalRule.getDiscountEntityType().equals(RuleDiscountEntityType.ITEM)) {
                ItemValueDiscount valueDiscount = (ItemValueDiscount) promotionalRule;
                itemPromotionalRuleMap.put(valueDiscount.getItemCode(), promotionalRule);
            }
        }
        return itemPromotionalRuleMap;
    }
}
